### ISSUE TITLE
Extraction logic is vulnerable to Zip Bomb attacks

### ISSUE THREAD

**Issue by zamazan4ik** [zip-rs/zip-old#228]
Hi!

Thank you for the library!

I have a question: how this library handles zip bombs. Is there any built-in protection? If no, can you suggest an option, how can I handle this?

My use case is: the user can upload a zip file into the system, then the system shall extract some information from the archive and store it somewhere. As you see, the user can try to send some malformed archives and crash my system.

Thanks!

**Comment by Plecra** [comment]
My understanding of zip bombs is that they specifically target antivirus software by appearing to contain zip files which contain zip files, and so on. This can crash an antivirus that attempts to recursively extract them, but shouldn't have an affect on any of `zip`s APIs as they are non recursive.

Are you aware of a method to attack the current implementation of `extract`?

**Comment by ZaMaZaN4iK** [comment]
No, I'm not aware of any kind of already existing method to attack the current implementation (actually I didn't try any).

About zip-bombs. Not only antivirus. E.g. recursive zip bombs can just cause DoS during the extraction process. Why I am explicitly asking about such things - because I remember already existing similar CVE in other ZIP libraries. E.g this one: https://bugs.python.org/issue36462 (yeah, that's not Rust and possibly that's attack is not applicable to zip-rs - just an example).

**Comment by Plecra** [comment]
Thanks for the link :) My first impression is a bit tangential to the `zip` library's threat model, and we provide the apis for a user to inspect the size of a zip file. Documenting the potential risk of writing untrusted data to the filesystem is a good idea.

For that, I'd like to have a clearer picture of the potential of this kind of attack...

- `compressed_size:uncompressed_size` ratio
- Duplicate file entries
- Anything else?

If we can reliably identify 'unreasonable' zip files, then I'd also be happy to include those checks in archive construction.

**Comment by Plecra** [comment]
To answer your initial question, there is no protection in place at the moment and you can defend against this kind of attack by setting reasonable upper limits on the sizes of the files in the archives your users upload. 

The only reason we don't do this by default is that it's trickier to determine what is 'reasonable' for all applications.

**Comment by ZaMaZaN4iK** [comment]
Thanks for the answer. Let's left the issue open. Maybe we will do some zip attacks analysis later (I'm not a professional in this area) and at least document some possible ways to prevent the most dangerous attack vectors.

**Comment by zamazan4ik** [comment]
Just found a PR with a bunch of Zip related security fixes: https://github.com/rust-compress/rc-zip/pull/18

**Comment by Plecra** [comment]
Interesting! This is good to see: We have introduced mitigations for some of these in the `enclosed_name` API, and the others depend on symlinks, which we are deliberately delaying support for due to these concerns.

**Comment by Plecra** [comment]
Thanks a lot for that reference! Knowing how they're setup will help a lot with these zip bombs.

I expect you're right that any overlapped files come from bad actors, in which case the mitigation in 0.6 can be part of construction - we can `files.sort(); files.dedup_by_key(|file| file.offset);`.

I'd be happy to quickly merge a PR with that change that passed CI

**Comment by aweinstock314** [comment]
I don't think that would suffice, since the CDH's don't have to point to consecutive LFH's.
![zip_nonconsecutive_overlap](https://user-images.githubusercontent.com/2462937/160248138-ad7865a2-7faf-4bcc-a808-68eb794e2424.png)

The article mentions that ClamAV's first fix had this same mistake:
```
In ClamAV bug 12356, Hanno Böck reported that zblg.zip caused high CPU usage in clamscan. An initial patch to detect overlapping files turned out to be incomplete because it only checked adjacent pairs of files.
```

**Comment by aweinstock314** [comment]
Actually, if you're only checking that the offsets in the CDH's are identical, you're only defending against exact overlap (![](https://www.bamsoftware.com/hacks/zipbomb/overlap.svg)) not partial overlap (![](https://www.bamsoftware.com/hacks/zipbomb/quote.svg)).

Detecting partial overlap would be something closer to
```rust
fn range_intersect<T: Ord>(x: Range<T>, y: Range<T>) -> Range<T> {
    Range {
        start: x.start.max(y.start),
        end: x.end.min(y.end),
    }
}

impl Shared {
    fn detect_overlap(&self) -> ZipResult<()>
        // covered[a] == b <-> an LFH spans [a, b)
        let mut covered: BTreeMap<u64, u64> = BTreeMap::new();
        for f in self.files.iter() {
            let a = f.header_start;
            let b = a + f.data_start.load() + f.compressed_size() + /* might need to account for data descriptor size? */;
            let x = a..b;
            // starting at 0 is quadratic in the number of files, is there a more precise interval query?
            // (a..b would not be accurate enough, consider the map containing 2..8, and trying to query 4..8)
            // I assumed BTreeMap supported "find the largest key smaller than a given key" in O(log(n)), but I didn't see it.
            // (It looks like JVM's TreeMap calls this operation `floorKey`)
            for (c, d) in covered.range(covered.floor_key(a)..b) {
                let y = (*c, *d);
                if !range_intersect(x, y).is_empty() {
                    return Err(ZipError::InvalidArchive("overlapping LFH's detected, possible zip bomb"));
                }
            }
            covered.insert(a, b);
        } 
        Ok(())
    }
}
```

**Comment by aweinstock314** [comment]
Ah, since `std::collections::btree_map::Range` implements `DoubleEndedIterator`, we can implement `floor_key` in `O(log(n))` as something like:
```rust
fn floor_key<K: Ord, V>(map: &'a BTreeMap<K, V>, k: &K) -> Option<&'a K> {
    map.range(0..=k).next_back().map(|(k, _)| k)
}
```

**Comment by Plecra** [comment]
Ah, that partial overlap attack is more troublesome :) Alright, I agree that we'll want the more involved checks. In my defence though I *did* add the `sort()` to address the bug you mentioned 😂

I think sorting by the start of the file should still work - the file list can be reduced to `Vec<(start: u64, len: u64)>` and

```rust
kernels.sort()
for i in 0..(kernels.len() - 1) {
    if kernels[i].start + kernels[i].len > kernels[i + 1].start { 
        boom();
    }
}
```

**Comment by Plecra** [comment]
I'd like to address the DOS vuln here as we discussed. Hope you don't mind me pinging you both once I've got an implementation to test!

**Comment by Pr0methean** [comment]
Sometimes overlapping files in a ZIP have a legitimate use (although it would still be useful to detect them in order to use https://crates.io/crates/reflink-copy). However, commit https://github.com/zip-rs/zip2/commit/aa890fa634785784bfce19d897828fa6c348556c will let you determine the total decompressed size of an archive before extracting (as long as it's written at the start of the file), or None if we're uncertain or if the total size is somehow more than u128::MAX bytes. This will be released in v1.2.0.

**Comment by Pr0methean** [comment]
If anyone still has concerns about this issue that the `decompressed_size()` function doesn't solve, please open a new issue at https://github.com/zip-rs/zip2/issues. This repo is no longer maintained.
